mesa/experimental/cell_space.py

```diff
@@ -0,0 +1,204 @@
+        return len(self.agents) == 0
+
+    @property
+    def is_full(self) -> bool:
+        """Returns a bool of the contents of a cell."""
+        return len(self.agents) == self.capacity
+
+    def __repr__(self):
+        return f"Cell({self.coordinate}, {self.agents})"
+
+    @cache
+    def neighborhood(self, radius=1, include_center=False):
+        if radius == 0:
+            return {self: self.agents}
+        if radius == 1:
+            return CellCollection(
+                {neighbor: neighbor.agents for neighbor in self._connections}
+            )
+        else:
+            neighborhood = {}
+            for neighbor in self._connections:
+                neighborhood.update(neighbor.neighorhood(radius - 1, include_center))
+            if not include_center:
+                neighborhood.pop(self, None)
+            return CellCollection(neighborhood)
+
+
+class CellCollection:
+    def __init__(self, cells: dict[Cell, list[Agent]] | Iterable[Cell]) -> None:
+        if isinstance(cells, dict):
+            self._cells = cells
+        else:
+            self._cells = {cell: cell.agents for cell in cells}
+
+    def __iter__(self):
+        return iter(self._cells)
+
+    def __getitem__(self, key):
+        return self._cells[key]
+
+    @cached_property
+    def __len__(self):
+        return len(self._cells)
+
+    def __repr__(self):
+        return f"CellCollection({self._cells})"
+
+    @cached_property
+    def cells(self):
+        return list(self._cells.keys())
+
+    @property
+    def agents(self):
+        return itertools.chain.from_iterable(self._cells.values())
+
+    def select_random_cell(self):
+        return random.choice(self.cells)
+
+    def select_random_agent(self):
+        return random.choice(list(self.agents))
+
+    def select(self, filter_func: Optional[Callable[[Cell], bool]] = None, n=0):
+        if filter_func is None and n == 0:
+            return self
+
+        return CellCollection(
+            {
+                cell: agents
+                for cell, agents in self._cells.items()
+                if filter_func is None or filter_func(cell)
+            }
+        )
+
+
+class Space:
+    cells: dict[Coordinate, Cell] = {}
+
+    def _connect_single_cell(self, cell):
+        ...
+
+    @cached_property
+    def all_cells(self):
+        return CellCollection({cell: cell.agents for cell in self.cells.values()})
+
+    def __iter__(self):
+        return iter(self.cells.values())
+
+    def __getitem__(self, key):
+        return self.cells[key]
+
+    def move_agent(self, agent: Agent, pos: Coordinate) -> None:
+        """Move an agent from its current position to a new position."""
+        self.cells[pos].add_agent(agent)
+
+    @property
+    def empties(self) -> CellCollection:
+        return self.all_cells.select(lambda cell: cell.is_empty)
+
+    def move_to_empty(self, agent: Agent) -> None:
+        while True:
+            new_cell = self.all_cells.select_random_cell()
```

Comment:
This is faster (in most cases) than getting the empty cells and selecting one of those I guess?

Would be interesting what happens with a 90% and 99% full grid.

Reply:
Exactly, this is usually the fastest option. But yes once you get an almost full grid it gets increasingly slow and if your grid is full it indeed runs forever. Thats why i added the (currently dead) code below, but it was much slower and didn't have the time yet to look into it more.

Reply:
The entire issue starts from not wanting a coupling between cells and the grid. In my version, I did allow for such coupling. Basically, cells update their owning grid when they are empty or when they are no longer empty. It makes move_to_empty in the grid simple because it can reuse the old optimized code. 

I understand @Corvince motivation (see https://github.com/projectmesa/mesa/issues/1900#issuecomment-1905592341) but I am not sure whether the price of this is worth it. Also, conceptually, I don't see how cells can exist without an "owning" DiscreteSpace.